core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushMergeWriterUpdateIntoConnector.java

```diff
@@ -0,0 +1,157 @@
+
+public class PushMergeWriterUpdateIntoConnector
+        implements Rule<TableFinishNode>
+{
+    private static final Capture<TableScanNode> TABLE_SCAN = newCapture();
+    private static final Capture<ProjectNode> PROJECT_NODE_CAPTURE = newCapture();
+    private static final Capture<MergeWriterNode> MERGE_WRITER_NODE_CAPTURE = newCapture();
+
+    private static final Pattern<TableFinishNode> PATTERN =
+            tableFinish().with(source().matching(
+                    mergeWriter().capturedAs(MERGE_WRITER_NODE_CAPTURE).with(source().matching(
+                            mergeProcessor().with(source().matching(
+                                    project().with(source().matching(
+                                            tableScan().capturedAs(TABLE_SCAN))).capturedAs(PROJECT_NODE_CAPTURE)))))));
+
+    public PushMergeWriterUpdateIntoConnector(PlannerContext plannerContext, TypeAnalyzer typeAnalyzer, Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.typeAnalyzer = requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+        this.plannerContext = requireNonNull(plannerContext, "plannerContext is null");
+    }
+
+    private final Metadata metadata;
+    private final PlannerContext plannerContext;
+    private final TypeAnalyzer typeAnalyzer;
+
+    @Override
+    public Pattern<TableFinishNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(TableFinishNode node, Captures captures, Context context)
+    {
+        TableScanNode tableScan = captures.get(TABLE_SCAN);
+        ProjectNode project = captures.get(PROJECT_NODE_CAPTURE);
+        MergeWriterNode mergeWriter = captures.get(MERGE_WRITER_NODE_CAPTURE);
+
+        List<String> orderedColumnNames = mergeWriter.getTarget().getMergeParadigmAndTypes().getColumnNames();
+        List<? extends Node> mergeAssignments = project.getAssignments().get(new Symbol("merge_row")).getChildren();
+
+        if (mergeAssignments.size() < orderedColumnNames.size()) {
+            return Result.empty();
+        }
+        ImmutableMap.Builder<String, Domain> assignmentsBuilder = ImmutableMap.builder();
+
+        for (int i = 0; i < orderedColumnNames.size(); i++) {
+            String columnName = orderedColumnNames.get(i);
+            Node assigmentNode = mergeAssignments.get(i);
+            checkForFailNode(assigmentNode);
+
+            ConnectorExpressionTranslator.ConnectorExpressionTranslation expressionTranslation = ConnectorExpressionTranslator.translateConjuncts(
+                    context.getSession(),
+                    ((Expression) assigmentNode),
+                    context.getSymbolAllocator().getTypes(),
+                    plannerContext,
+                    typeAnalyzer);
+
+            ConnectorExpression connectorExpression = expressionTranslation.connectorExpression();
+
+            if (connectorExpression instanceof Constant constantExpression) {
```

Comment:
Shall we stop here if any expression not the `Constant`, are we allow only push the partial parts in the update statement?

Reply:
Looks, like here we are iterating over all parts of merge row, so some of them could be not constants, but we are skipping them, so we try to find all constants, if not then we fall back to default update:
```
 if (assignments.isEmpty()) {
            return Result.empty();
        }
```

Reply:
I don't understand this logic. Seems like the `else` branch is missing as right now we are skipping expressions that are non-constants. Should we return Result.empty in that case?

Reply:
Here we iterating through merge assignments of our merge row and collecting only Constants (it means direct updates, like set some column to 100, we don't yet support more clever expression like set some_column = some_column + 1)

After we iterate all over our assignments and collected constant Update to assignments map, 
we are checking is this map empty or not. If empty - it means that we don't have any constant updates and we returning empty result, else proceed with next logic

Reply:
https://github.com/trinodb/trino/pull/16445#discussion_r1168590968
here I described a little be how merge_row and merge assignments looks

Reply:
What about the situation when we have both constant and non-constant assignments `SET value = 1, value2 = value2 + 1` ? Should we verify that ALL of the assignments are constant?

Reply:
Thanks, good catch I need to add such check

Reply:
Check added in case we have` value2 = value2 + 1` - it's transformed to function CALL

Reply:
@wendigo I decided to simplify this first iteration as much as possible, so I removed most of the checks like checkForFailNode e.t.c the only check that I left - it's check that in merge row ALL our merge assignments are constant update or Symbols (column untouched). In all other cases we fall back to default path and return empty result.